src/Parallel/Actions/SetupDataBox.hpp

```diff
@@ -0,0 +1,132 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <tuple>
+#include <type_traits>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "ParallelAlgorithms/Initialization/MergeIntoDataBox.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace Actions {
+/// \cond
+struct SetupDataBox;
+/// \endcond
+
+namespace detail {
+CREATE_HAS_TYPE_ALIAS(simple_tags)
+CREATE_HAS_TYPE_ALIAS_V(simple_tags)
+CREATE_HAS_TYPE_ALIAS(compute_tags)
+CREATE_HAS_TYPE_ALIAS_V(compute_tags)
+
+template <typename Action, typename enable = std::bool_constant<true>>
+struct optional_simple_tags;
+
+template <typename Action>
+struct optional_simple_tags<
+    Action, std::bool_constant<detail::has_simple_tags_v<Action>>> {
+  using type = typename Action::simple_tags;
+};
+
+template <typename Action>
+struct optional_simple_tags<
+    Action, std::bool_constant<not detail::has_simple_tags_v<Action>>> {
+  using type = tmpl::list<>;
+};
+
+template <typename Action, typename enable = std::bool_constant<true>>
+struct optional_compute_tags;
+
+template <typename Action>
+struct optional_compute_tags<
+    Action, std::bool_constant<detail::has_compute_tags_v<Action>>> {
+  using type = typename Action::compute_tags;
+};
+
+template <typename Action>
+struct optional_compute_tags<
+    Action, std::bool_constant<not detail::has_compute_tags_v<Action>>> {
+  using type = tmpl::list<>;
+};
+
+template <typename ActionList>
+struct get_action_list_simple_tags;
+
+template <typename... ActionList>
+struct get_action_list_simple_tags<tmpl::list<ActionList...>> {
+  using type = tmpl::remove_duplicates<tmpl::flatten<
+      tmpl::list<typename optional_simple_tags<ActionList>::type...>>>;
+};
+
+template <typename ActionList>
+struct get_action_list_compute_tags;
+
+template <typename... ActionList>
+struct get_action_list_compute_tags<tmpl::list<ActionList...>> {
+  using type = tmpl::remove_duplicates<tmpl::flatten<
+      tmpl::list<typename optional_compute_tags<ActionList>::type...>>>;
+};
+
+template <typename DbTags, typename... simple_tags, typename... compute_tags>
+static auto merge_into_databox_helper(
+    db::DataBox<DbTags>&& box, tmpl::list<simple_tags...> /*meta*/,
+    tmpl::list<compute_tags...> /*meta*/) noexcept {
+  if constexpr (not tmpl2::flat_any_v<
+                    tmpl::list_contains_v<DbTags, simple_tags>...> and
+                not tmpl2::flat_any_v<
+                    tmpl::list_contains_v<DbTags, compute_tags>...>) {
```

Comment:
I think the check that none of the compute tags already exist in the DataBox is not necessary, since `db::create_from` merges the compute tags anyway. The check makes the `SetupDataBox` action a bit more restrictive than it needs to be, so I suggest you remove it.

Reply:
ping, in case you missed this

Reply:
ah, sorry, forgot to comment -- this didn't work. My understanding of the error I encountered was an edge case associated with the base tags of the compute tags and logic involving multiple passes through the initialization phase (which we don't usually use, but must always compile correctly because the initialization phase still 'loops').

Reply:
Do you remember what specifically didn't work? Adding a compute tag multiple times should work. The main reason I'm suggesting to remove this restriction is that it can break in some edge-cases, e.g. when the `SetupDataBox` is not the very first action and the DataBox already has some compute tags in it for whatever reason.

Reply:
I think you must be recalling functionality that has since been removed in more recent `DataBox` changes. In particular, on develop the following patch that uses `db::create_from` with a new compute tag causes compilation failure:
```
diff --git a/tests/Unit/DataStructures/DataBox/Test_DataBox.cpp b/tests/Unit/DataStructures/DataBox/Test_DataBox.cpp
index da4a61866..955a5bf49 100644
--- a/tests/Unit/DataStructures/DataBox/Test_DataBox.cpp
+++ b/tests/Unit/DataStructures/DataBox/Test_DataBox.cpp
@@ -237,7 +237,11 @@ void test_databox() noexcept {
                            test_databox_tags::Lambda1Compute>>(
         3.14, std::vector<double>{8.7, 93.2, 84.7}, "My Sample String"s);
     /// [create_databox]
-    return original_box;
+    auto new_box =
+        db::create_from<db::RemoveTags<>, db::AddSimpleTags<>,
+                        db::AddComputeTags<test_databox_tags::Tag4Compute>>(
+            std::move(original_box));
+    return new_box;
   };
   {
     const auto original_box = create_original_box();
```
Or perhaps are you thinking of the `initialization::merge_into_databox` utility? I didn't use that because it is more restrictive in requiring equivalence operators for the tagged types.

Reply:
You're right, only `merge_into_databox` resolves duplicate compute items. I suggest you do the same here, adding only compute items that are not already present in the DataBox. This is the relevant snippet from `MergeIntoDataBox.hpp`:

```cpp
  using compute_tags_to_add = tmpl::remove_if<
      ComputeTagsList,
      tmpl::bind<tmpl::list_contains,
                 tmpl::pin<typename db::DataBox<DbTagsList>::compute_item_tags>,
                 tmpl::_1>>;
```

Reply:
I'm not pushing for the functionality change, but I want to point out that this is more simply written as `tmpl::list_difference<ComputeTagsList, typename db::DataBox<DbTagsList>::compute_item_tags>`.

Reply:
I've added another fixup that performs the change for both compute tags and for simple tags -- now both have 'merge' behavior, where they'll just add tags that aren't already present.